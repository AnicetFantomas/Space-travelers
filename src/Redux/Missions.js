import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';

const baseUrl = 'https://api.spacexdata.com/v3/missions';

export const fetchMissionData = createAsyncThunk(
  'action/fetchMissionData',
  async () => {
    const res = await fetch(baseUrl);
    const data = await res.json();
    const extractData = data.map((mission) => ({
      id: mission.mission_id,
      name: mission.mission_name,
      description: mission.description,
      manufacturer: mission.manufacturers[0],
      wiki: mission.wikipedia,
      joined: false,
    }));
    return extractData;
  },
);

export const missionSlice = createSlice({
  name: 'reduce/fetchMissionData',
  initialState: [],
  reducers: {
    updateMissionState: (state, action) => state.map((mission) => {
      if (mission.id === action.payload) {
        return { ...mission, joined: !mission.joined };
      }
      return mission;
    }),
  },
  extraReducers: {
    [fetchMissionData.fulfilled]: (state, action) => action.payload,
  },
});

export const { updateMissionState } = missionSlice.actions;

export default missionSlice.reducer;
