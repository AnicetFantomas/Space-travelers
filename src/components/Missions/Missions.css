table {
  width: 100%;
  border-collapse: collapse;
  background-color: #fff;
}

tr {
  width: 100%;
}

th {
  font-weight: 500;
  font-size: 1.2rem;
}

td,
th {
  border: 1px solid #ddd;
  text-align: left;
  padding: 8px;
}

td:first-child {
  font-weight: bold;
}

tr:nth-child(even) {
  background-color: #cac9c9;
}

button {
  padding: 0.5rem;
  border-radius: 7px;
  width: max-content;
}

.inactive {
  background-color: #c2c1c1;
}

.active {
  background-color: #4386bf;
  color: #fff;
}

.join {
  border: 2px solid green;
  color: green;
  background: none;
}

.leave {
  border: 2px solid red;
  color: red;
  background: none;
}
