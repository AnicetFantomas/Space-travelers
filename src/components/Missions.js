import React, { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { fetchMissionData, updateMissionState } from '../Redux/Missions';

const Missions = () => {
  const getAllMissions = useSelector((state) => state.missions);
  const dispatch = useDispatch();

  useEffect(() => {
    if (!getAllMissions.length) dispatch(fetchMissionData());
  }, [dispatch]);

  const handleClick = ({ target }) => {
    const { id } = target;
    dispatch(updateMissionState(id));
  };

  const activeButton = <button type="button">Active memeber</button>;
  const inactiveButton = <button type="button">Not a member</button>;

  return (
    <table>
      <thead>
        <tr>
          <th>Mission</th>
          <th>description</th>
          <th>Status</th>
          <th aria-label="Empty cell" />
        </tr>
      </thead>
      <tbody>
        {getAllMissions.map((mission) => (mission.joined ? (
          <tr key={mission.id}>
            <td>{mission.name}</td>
            <td>{mission.description}</td>
            <td>{activeButton}</td>
            <td><button id={mission.id} type="button" onClick={handleClick}>Leave Mission</button></td>
          </tr>
        )
          : (
            <tr key={mission.id}>
              <td>{mission.name}</td>
              <td>{mission.description}</td>
              <td>{inactiveButton}</td>
              <td><button id={mission.id} type="button" onClick={handleClick}>Join Mission</button></td>
            </tr>
          )))}
      </tbody>
    </table>
  );
};
export default Missions;
