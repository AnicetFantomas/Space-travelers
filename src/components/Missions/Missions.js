import React, { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import './Missions.css';
import { fetchMissionData, updateMissionState } from '../../Redux/Missions';

const Missions = () => {
  const getAllMissions = useSelector((state) => state.missions);
  const dispatch = useDispatch();

  useEffect(() => {
    if (!getAllMissions.length) dispatch(fetchMissionData());
  }, [dispatch]);

  const handleClick = ({ target }) => {
    const { id } = target;
    dispatch(updateMissionState(id));
  };

  const activeButton = <button type="button" className="active">Active memeber</button>;
  const inactiveButton = <button type="button" className="inactive">Not a member</button>;

  return (
    <table>
      <thead>
        <tr>
          <th>Mission</th>
          <th>description</th>
          <th>Status</th>
          <th aria-label="Empty cell" />
        </tr>
      </thead>
      <tbody>
        {getAllMissions.map((mission) => (
          <tr key={mission.id}>
            <td>{mission.name}</td>
            <td>{mission.description}</td>
            <td>
              {mission.joined && activeButton}
              {!mission.joined && inactiveButton}
            </td>
            <td>
              {mission.joined && <button id={mission.id} className="join" type="button" onClick={handleClick}>Leave Mission</button>}
              {!mission.joined && <button id={mission.id} className="leave" type="button" onClick={handleClick}>Join Mission</button>}
            </td>
          </tr>
        ))}
      </tbody>
    </table>
  );
};
export default Missions;
