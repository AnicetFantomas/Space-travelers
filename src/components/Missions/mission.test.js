import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import store from '../Navigation/configuration';
import Missions from './Missions';

describe('test missions component', () => {
    it('Render missions row', () => {
        const app = renderer
        .create(
            <Provider store={store}>
                <Missions />
            </Provider>
        ).toJSON();
        expect(app).toMatchSnapshot()
    })
})