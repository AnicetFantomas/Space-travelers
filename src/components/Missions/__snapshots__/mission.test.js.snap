// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test missions component Render missions row 1`] = `
<table>
  <thead>
    <tr>
      <th>
        Mission
      </th>
      <th>
        description
      </th>
      <th>
        Status
      </th>
      <th
        aria-label="Empty cell"
      />
    </tr>
  </thead>
  <tbody />
</table>
`;
